### ISSUE TITLE
[java] GuardLogStatement: Documentation is unclear why getters are flagged

### ISSUE THREAD

**Issue by soloturn** [pmd/pmd#4731]
**Affects PMD Version:** 
7.0.0-rc4

**Rule:**
[GuardLogStatement](https://docs.pmd-code.org/pmd-doc-7.0.0-rc4/pmd_rules_java_bestpractices.html#guardlogstatement): Logger calls should be surrounded by log level guards.

**Description:**
```java
            if (handler.isPresent()) {
                handlerMap.put(field, handler.get());
            } else {
                logger.error("Unsupported field: '{}.{}'", type.getId(), field.getName());
            }
        }

```

other example:
```java
            if (result == null) {
                logger.warn("Unknown enum value: '{}' for enum {}", data.getAsString(), enumType.getSimpleName());
            }

```

**Expected outcome:**

PMD reports a violation at line "logger.error", the example in the documentation says: no guard necessary if parameters are there; getters are parameters.

**Running PMD through:** *Gradle*

**Comment by soloturn** [comment]
saw a solution to it here, applied, and it works, with 7.0.0-rc4.
https://stackoverflow.com/questions/35096746/substituting-parameters-in-log-message-and-add-a-throwable-in-log4j-2

**Comment by jsotuyod** [comment]
There is no problem with the `if` itself.  You can remove the conditional and this still gets flagged.

The current rule will flag this because of the method invocations (getters), as these imply that work is actually performed even if the logger is disabled at the given level.

Moreover, the methods *may* have side-effects, with the code reliant on these, but that's just a bonus of having this rule around, not the actual intention.

We could consider adding a property to whitelist getters, but seems more like an enhancement than a FP under the rule's current guise.

**Comment by jsotuyod** [comment]
Upon rethinking about this, I don't think this property makes much sense, unless targeting Java < 8… any getter can be immediately turned into a method reference to make it's invocation completely lazy.

In your examples, it would suffice to do:

```java
if (handler.isPresent()) {
    handlerMap.put(field, handler::get);
} else {
    logger.error("Unsupported field: '{}.{}'", type::getId, field::getName);
}
```

```java
if (result == null) {
    logger.warn("Unknown enum value: '{}' for enum {}", data::getAsString, enumType::getSimpleName);
}
```

Just making this explicitly on the documentation should be enough…
